# misc-test.v01
Test
1. toDataFame
1. toDateParts

## General

In [1]:
__author__ = 'BruceCottman'
__author__

In [2]:
!conda list

# packages in environment at /Users/brucecottman/anaconda3/envs/steppy:
#
# Name                    Version                   Build  Channel
_ipyw_jlab_nb_ext_conf    0.1.0            py36h2fc01ae_0  
alabaster                 0.7.10           py36h174008c_0  
anaconda                  5.2.0                    py36_3  
anaconda-client           1.6.14                   py36_0  
anaconda-navigator        1.8.7                    py36_0  
anaconda-project          0.8.2            py36h9ee5d53_0  
appnope                   0.1.0            py36hf537a9a_0  
appscript                 1.0.1            py36h9e71e49_1  
asn1crypto                0.24.0                   py36_0  
astroid                   1.6.3                    py36_0  
astropy                   3.0.2            py36h917ab60_1  
atomicwrites              1.2.1                     <pip>
attrs                     18.1.0                   py36_0  
Augmentor                 0.2.2                     <pip>
babel                  

## imports

In [3]:
import os,sys
sys.path.append(os.path.abspath('../'))
sys.path

['',
 '/Users/brucecottman/anaconda3/envs/steppy/lib/python36.zip',
 '/Users/brucecottman/anaconda3/envs/steppy/lib/python3.6',
 '/Users/brucecottman/anaconda3/envs/steppy/lib/python3.6/lib-dynload',
 '/Users/brucecottman/.local/lib/python3.6/site-packages',
 '/Users/brucecottman/anaconda3/envs/steppy/lib/python3.6/site-packages',
 '/Users/brucecottman/anaconda3/envs/steppy/lib/python3.6/site-packages/aeosa',
 '/Users/brucecottman/anaconda3/envs/steppy/lib/python3.6/site-packages/pip-18.0-py3.6.egg',
 '/Users/brucecottman/anaconda3/envs/steppy/lib/python3.6/site-packages/IPython/extensions',
 '/Users/brucecottman/.ipython',
 '/Users/brucecottman/Documents/PROJECTS/steppy-toolkit']

In [4]:
import pandas as pd
import copy, random
import numpy as np
from numba import jit

In [5]:
import steppy
from steppy.base import Step, BaseTransformer, make_transformer
from steppy.adapter import Adapter, E
from steppy.utils import get_logger
import joblib

In [6]:
from toolkit.preprocessing.S_toDataFrame import S_toDataFrame

In [7]:
#from S_misc import S_toDataFrame


Level	Numeric value
-   CRITICAL	50
-    ERROR	40
-    WARNING	30
-    INFO	20
-    DEBUG	10
-    NOTSET	0

In [8]:
!conda list

# packages in environment at /Users/brucecottman/anaconda3/envs/steppy:
#
# Name                    Version                   Build  Channel
_ipyw_jlab_nb_ext_conf    0.1.0            py36h2fc01ae_0  
alabaster                 0.7.10           py36h174008c_0  
anaconda                  5.2.0                    py36_3  
anaconda-client           1.6.14                   py36_0  
anaconda-navigator        1.8.7                    py36_0  
anaconda-project          0.8.2            py36h9ee5d53_0  
appnope                   0.1.0            py36hf537a9a_0  
appscript                 1.0.1            py36h9e71e49_1  
asn1crypto                0.24.0                   py36_0  
astroid                   1.6.3                    py36_0  
astropy                   3.0.2            py36h917ab60_1  
atomicwrites              1.2.1                     <pip>
attrs                     18.1.0                   py36_0  
Augmentor                 0.2.2                     <pip>
babel                  

qtconsole                 4.3.1            py36hd96c0ff_0  
qtpy                      1.4.1                    py36_0  
readline                  7.0                  hc1231fa_4  
requests                  2.18.4           py36h4516966_1  
retrying                  1.3.3                     <pip>
rope                      0.10.7           py36h68959ac_0  
ruamel_yaml               0.15.35          py36h1de35cc_1  
scikit-image              0.13.1           py36h1de35cc_1  
scikit-learn              0.19.1           py36hffbff8c_0  
scipy                     1.1.0            py36hcaad992_0  
seaborn                   0.8.1            py36h595ecd9_0  
send2trash                1.5.0                    py36_0  
setuptools                39.1.0                   py36_0  
setuptools                40.2.0                    <pip>
simplegeneric             0.8.1                    py36_2  
singledispatch            3.4.0.3          py36hf20db9d_0  
sip                       4.19.8           p

In [9]:
from __future__ import print_function
import matplotlib as mpl
from matplotlib import pyplot as plt
from matplotlib import cm
import probscale
import seaborn as sns
from scipy.stats import probplot
import statsmodels.api as sm

%reload_ext autoreload
%autoreload 2
%matplotlib inline
import matplotlib


## logging

In [10]:
logger = get_logger()
logger.setLevel('DEBUG'.upper())
logger.getEffectiveLevel()

10

## Clean out EXPERIMENT_DIR

In [11]:
MODELS_DIR = '../input/models'
EXPERIMENT_DIR = MODELS_DIR+'/tmp'
!mkdir $MODELS_DIR
!mkdir $EXPERIMENT_DIR
import shutil

# By default pipelines will try to load previously trained models so we delete the cache to ba sure we're starting from scratch
shutil.rmtree(EXPERIMENT_DIR, ignore_errors=True)

mkdir: ../input/models: File exists
mkdir: ../input/models/tmp: File exists


## Input  Datasets

Continuous features can exhibit a wide range of distributional characteristics, including near-Gaussian, strongly asymmetric, light- or heavy-tailed, multi-moda

Catagorical features  can be  either ordered, or unordered. Catagorical features are not covered here.

### Boston Dataset

In [12]:
from sklearn.datasets import load_boston
boston = load_boston()

logger.info(boston.feature_names)
logger.info(boston.DESCR)
City = pd.DataFrame(boston.data, columns = boston.feature_names )
City = City[['CRIM', 'INDUS','NOX','TAX','B']]
City['MEDV'] = boston.target
logger.info(City.shape)
logger.info(City.columns)

2018-09-07 20:04:40 steppy >>> ['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']
2018-09-07 20:04:40 steppy >>> Boston House Prices dataset

Notes
------
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index 

A dataset that illustrates the range of distributional behavior is the Boston Boston dataframe.  This dataframe includes 14 numerical variables that characterize 506 suburban housing tracts in the Boston area: 12 of these variables have class “numeric” and the remaining two have class “integer”.  

-  the distribution of crim is very asymmetric with an extremely heavy right tail;
- the distribution of lstat is also clearly asymmetric but far less so;
- the distribution of rm appears to be almost Gaussian; 
- the distribution of zn appears to be tri-modal, mostly concentrated around zero, but with clear secondary peaks at around 20 and 80.

### Housing Dataset

- Feature 0 (median income in a block) and feature 5 (number of households) of
the `California housing dataset
<http://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html>`_ have very
different scales and contain some very large outliers. These two
characteristics lead to difficulties to visualize the data and, more
importantly, they can degrade the predictive performance of many machine
learning algorithms. Unscaled data can also slow down or even prevent the
convergence of many gradient-based estimators.
- Indeed many estimators are designed with the assumption that each feature takes
values close to zero or more importantly that all features vary on comparable
scales. In particular, metric-based and gradient-based estimators often assume
approximately standardized data (centered features with unit variances). A
notable exception are decision tree-based estimators that are robust to
arbitrary scaling of the data.

In [13]:
from sklearn.datasets import fetch_california_housing
print(__doc__)
dataset = fetch_california_housing()
X_full, y_full = dataset.data, dataset.target

# Take only 2 features to make visualization easier
# Feature of 0 has a long tail distribution.
# Feature 5 has a few but very large outliers.
logger.info(dataset.data.shape)
logger.info(dataset.feature_names)
logger.info(dataset.DESCR) 
Housing = pd.DataFrame(dataset.data, columns = dataset.feature_names )
keeppers = ['MedInc', 'HouseAge', 'AveRooms'
                   , 'AveBedrms', 'Population', 'AveOccup']
Housing = Housing[keeppers]
logger.info(Housing.shape)
logger.info(Housing.columns)

Automatically created module for IPython interactive environment
2018-09-07 20:04:40 steppy >>> (20640, 8)
2018-09-07 20:04:40 steppy >>> ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']
2018-09-07 20:04:40 steppy >>> California housing dataset.

The original database is available from StatLib

    http://lib.stat.cmu.edu/datasets/

The data contains 20,640 observations on 9 variables.

This dataset contains the average house value as target variable
and the following input variables (features): average income,
housing average age, average rooms, average bedrooms, population,
average occupation, latitude, and longitude in that order.

References
----------

Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions,
Statistics and Probability Letters, 33 (1997) 291-297.


2018-09-07 20:04:40 steppy >>> (20640, 6)
2018-09-07 20:04:40 steppy >>> Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population',
       'AveOccup'],


### Hip_m Dataset

In [14]:
'''
hip_m=np.genfromtxt('HIP_MEN_chr_pos_rs_pval.txt', dtype=None)
hip_m= pd.DataFrame(hip_m[['f0','f1','f2','f3']])
Hip_m= hip_m
hip_m= hip_m[['f1','f3']]
logger.info(hip_m.shape)
logger.info(hip_m.head())
'''


"\nhip_m=np.genfromtxt('HIP_MEN_chr_pos_rs_pval.txt', dtype=None)\nhip_m= pd.DataFrame(hip_m[['f0','f1','f2','f3']])\nHip_m= hip_m\nhip_m= hip_m[['f1','f3']]\nlogger.info(hip_m.shape)\nlogger.info(hip_m.head())\n"

# toDataFrame

## Adaptors

In [15]:
some_other_bound_variable = 0
Housing = Housing.values
Housing_o = Housing.copy()

In [16]:
data_Housing_str = {
    'input': {
        'features': Housing,
        'features_o': Housing_o,
        'labels': 'Income'
    }
}
data_Housing_list = {
    'input': {
        'features': Housing,
        'features_o': Housing_o,
        'labels': ['Income']
    }
}
data_Housing_mlist = {
    'input': {
        'features': Housing,
        'features_o': Housing_o,
        'labels': keeppers +['bogie']
    }
}
data_Housing_bad = {
    'input': {
        'features': np.ndarray([1,2,3]),
        'features_o':  np.ndarray([1,2,3]),
        'labels': 'Income'
    }
}
City_o = City.copy()
data_City = {
    'input': {
        'features': City,
        'features_o': City_o,
        'labels': ['CRIME']
    }
}

'''
Hip_m_o = Hip_m.copy()
data_Hip_m = {
    'input': {
        'features': Hip_m,
        'features_o': Hip_m_o
    }
}
'''

"\nHip_m_o = Hip_m.copy()\ndata_Hip_m = {\n    'input': {\n        'features': Hip_m,\n        'features_o': Hip_m_o\n    }\n}\n"

## toDataFrame_step

In [17]:
toDataFrame_step = Step(name='toDataFrame',
               transformer=S_toDataFrame(),
               input_data=['input'],
               adapter=Adapter({
                   'X': E('input','features_o'),
                    'labels': E('input','labels')
               }),
               experiment_directory=EXPERIMENT_DIR,
                is_trainable=False,
                force_fitting=False)

2018-09-07 20:04:40 steppy >>> initializing Step toDataFrame...
2018-09-07 20:04:40 steppy >>> initializing experiment directories under ../input/models/tmp
2018-09-07 20:04:40 steppy >>> done: initializing experiment directories
2018-09-07 20:04:40 steppy >>> Step toDataFrame initialized


In [18]:
toDataFrame_step 

InvocationException: GraphViz's executables not found

## toDataFrame_step test

In [24]:
#causes error
try:
    print(toDataFrame_step.transform(data_Housing_bad)['X'].head(2))
except: pass

2018-09-07 20:05:25 steppy >>> Step toDataFrame, adapting inputs...
2018-09-07 20:05:25 steppy >>> Step toDataFrame, transforming...


In [25]:
toDataFrame_step.transform(data_Housing_str)['X'].head(2)

2018-09-07 20:05:25 steppy >>> Step toDataFrame, adapting inputs...
2018-09-07 20:05:25 steppy >>> Step toDataFrame, transforming...


Income  Column_1  Column_2  Column_3  Column_4  Column_5
0  8.3252      41.0  6.984127   1.02381     322.0  2.555556
1  8.3014      21.0  6.238137   0.97188    2401.0  2.109842

In [26]:
toDataFrame_step.transform(data_Housing_list)['X'].head(2)

2018-09-07 20:05:25 steppy >>> Step toDataFrame, adapting inputs...
2018-09-07 20:05:25 steppy >>> Step toDataFrame, transforming...


Income  Column_1  Column_2  Column_3  Column_4  Column_5
0  8.3252      41.0  6.984127   1.02381     322.0  2.555556
1  8.3014      21.0  6.238137   0.97188    2401.0  2.109842

In [27]:
toDataFrame_step.transform(data_Housing_mlist)['X'].head(2)

2018-09-07 20:05:25 steppy >>> Step toDataFrame, adapting inputs...
2018-09-07 20:05:25 steppy >>> Step toDataFrame, transforming...


MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup
0  8.3252      41.0  6.984127    1.02381       322.0  2.555556
1  8.3014      21.0  6.238137    0.97188      2401.0  2.109842

In [28]:
toDataFrame_step.transform(data_City)['X'].head(2)

2018-09-07 20:05:26 steppy >>> Step toDataFrame, adapting inputs...
2018-09-07 20:05:26 steppy >>> Step toDataFrame, transforming...


CRIM  INDUS    NOX    TAX      B  MEDV
0  0.00632   2.31  0.538  296.0  396.9  24.0
1  0.02731   7.07  0.469  242.0  396.9  21.6

# References